Add rebar 3.7.x compatibility

nalundgaard opened the pull request:
* Add compatibility with new resource format introduced in rebar 3.7.0 in backwards-compatible way.

daniel-zinov committed:
compatiblity with new resource format

nalundgaard committed:
make absorb_resources compatible with all resource formats

* Move code handling rebar v2 resources to dedicated head and check it more reliably (`is_rec_type` check)
* Restore old format check for "legacy" resources (when using rebar3 <3.7.0)

zdaniel86 commented:
my changes are not sufficient to resolve the issue. This requires much more changes :( . I haven't managed to finish it.

nalundgaard committed:
Implement resource v2 callback compatibiity

Implement rebar_resource_v2 compatible callbacks in order to make the plugin compatible with all known versions of rebar3, following the guide at https://www.rebar3.org/v3/docs/custom-dep-resources#section-resources-compatible-with-all-versions

nalundgaard commented:
@zdaniel86 I think I just about have it? It looks like it's working now. Could you please have a look too?

rmpalomino commented on src/rebar_raw_resource.erl:
```diff
@@ -254,42 +272,72 @@ download(Dest, {?RTYPE, Spec, Opts}, State) ->
 %
-lock(Path, {?RTYPE, Loc, #mod_ref{res = Res, ref = Prev} = Rec}) ->
+%% pre-rebar3 3.7.0 format
+lock(Path, Source) when not(erlang:is_tuple(Path)) ->
```
The guard in the rebar3 docs checks that Source is a tuple instead of checking that Path is not a tuple, should we do that instead?

nalundgaard replied:
Yeah, I used this guard because the one in their example is useless, because it guards on source being a tuple, but AFAICT it's a tuple in the old style and the new style, so... yeah. This one actually verifies that the first arg isn't a record (which is the new first arg in the current resource format).

nalundgaard commented:
Looks like I need to put this on hold. While it doesn't categorically break with this change, it looks like all of the passthrough callbacks need to be fixed in order for this to actually function as expected. What a nightmare.

sg2342 commented:
sg2342/rebar_raw_resource@72fed2302eb7c64eea8147f62c2c1d287d120536 has a "works for me" solution for this problem.

nalundgaard commented:
@sg2342 looks exactly what I’ve been meaning to implement, seems like it should work. What I didn’t realize on my first stab was that you have to keep the old format callbacks separate to pass through to the old format modules in rebar < 3.7, and keep the args for the new format in the new callback pass throughs. Thanks for the tip! I’ll incorporate that into this PR.

nalundgaard commented:
@zdaniel86 @sg2342 @rmpalomino @motobob I *think* this is good to go now. Thanks very much to @sg2342 for the "kludge" hint (sg2342/rebar_raw_resource@72fed23); It almost worked for me. It writes a wrong lockfile, though (does not lock it as a raw file, so applying the lock, e.g. if you nuke `_build`, bypasses the plugin and results in a broken dep). So, I had to patch that to restore correct locking, and now it seems to be working for me.

nalundgaard commented on src/rebar_raw_resource.erl:
```diff
@@ -254,13 +280,60 @@ download(Dest, {?RTYPE, Spec, Opts}, State) ->
+                    % because download has not put the downloaded app into
+                    % its final location, configure the app dir to be the
+                    % Dest provided.
+                    TmpAppInfo = rebar_app_info:dir(AppInfo, Dest),
```
@sg2342 You might be interested in this. I found that in the case of an initial download, especially in the case where a lock is set but the _build dir has been nuked (either in total or selectively), because the download hasn't been relocated to the `_build/...` dir yet, you have to fake out the call to `Mod:make_vsn(AppInfo, [])` with a `dir` app_info setting of `Dest` (where the dependency is is downloaded before rebar3 later moves it).

nalundgaard commented on src/rebar_raw_resource.erl:
```diff
@@ -348,16 +476,19 @@ format_error(Mod, Func, Class, Data) ->
         -> mod_data() | no_return().
 absorb_app_infos(Data, [AppInfo | AppInfos]) ->
+    % absorb deps from app_info to ensure that all deps are absorbed
+    AppInfoDeps = rebar_app_info:get(AppInfo, deps, []),
```
This is another key call-out. I found that when there is a "raw" dependency of a dependency in a top-level application, it's often not included in the "deps" or "lock" rebar state. This can trigger an `unmapped_dep_loc` or `unmapped_dep_name` error when downloading a missing raw dep because of the rebar state. This call collects dependencies of dependencies, which appears to resolve the problem. If there is a more effective way to reliablty collect all the app specifications, that would be nice.

jadeallenx replied:
I think you should add a pointer to this PR somewhere in the comments for this function.

nalundgaard replied:
@mrallen1 I actually ended up folding and taking this back out. I've realized that there's actually no real need to save the `#mod_dep{}` records at all with `rebar_resource_v2` format, since the rebar_app_info contains all the data you need to do what's called for.

It really is the case that we could delete ~75% of the code in this repo if we didn't need to support pre-3.7 rebar3.

rmpalomino reviewed (approved):
This appears to be working for me after the latest update.
